examples/src/test/java/io/grpc/examples/routeguide/RouteGuideClientTest.java

```diff
@@ -0,0 +1,790 @@
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * Unit tests for {@link RouteGuideClient}.
+ * For demonstrating how to write gRPC unit test only.
+ * Not intended to provide a high code coverage or to test every major usecase.
+ *
+ * <p>For basic unit test examples see {@link io.grpc.examples.helloworld.HelloWorldClientTest} and
+ * {@link io.grpc.examples.helloworld.HelloWorldServerTest}.
+ */
+@RunWith(JUnit4.class)
+public class RouteGuideClientTest {
+  private final RouteGuideGrpc.RouteGuideImplBase serviceImpl =
+      spy(new RouteGuideGrpc.RouteGuideImplBase() {});
+
+  private Server fakeServer;
+  private RouteGuideClient client;
+
+  @Before
+  public void setUp() throws IOException {
+    String uniqueServerName = "fake server for " + getClass();
+    fakeServer =
+        InProcessServerBuilder.forName(uniqueServerName).addService(serviceImpl).build().start();
+    ManagedChannelBuilder channelBuilder = InProcessChannelBuilder.forName(uniqueServerName);
+    client = spy(
+        new RouteGuideClient(channelBuilder) {
+          int idx = 0;
+          @Override
+          int randomIndex(int range, Random random) {
+            // deterministic and incremental
+            return idx++ % range;
+          }
+
+          @Override
+          void sleep(long millis) {
+            // no sleep
+          }
+        });
+  }
+
+  @After
+  public void tearDown() throws InterruptedException {
+    if (client != null) {
+      client.shutdown();
+    }
+    if (fakeServer != null) {
+      fakeServer.shutdownNow();
+    }
+  }
+
+  /**
+   * Example for testing blocking unary call.
+   */
+  @Test
+  public void testGetFeature_exist() {
+    Point requestPoint =  Point.newBuilder().setLatitude(-1).setLongitude(-1).build();
+    Point responsePoint = Point.newBuilder().setLatitude(-123).setLongitude(-123).build();
+    final Feature responseFeature =
+        Feature.newBuilder().setName("dummyFeature").setLocation(responsePoint).build();
+    Answer<Void> answer =
+        new Answer<Void>() {
+          @SuppressWarnings("unchecked")
+          @Override
+          public Void answer(InvocationOnMock invocation) {
+            StreamObserver<Feature> responseObserver =
+                (StreamObserver<Feature>) invocation.getArguments()[1];
+            responseObserver.onNext(responseFeature);
+            responseObserver.onCompleted();
+            return null;
+          }
+        };
+    doAnswer(answer)
+        .when(serviceImpl)
+        .getFeature(any(Point.class), Matchers.<StreamObserver<Feature>>any());
+
+    client.getFeature(-1, -1);
+
+    verify(serviceImpl, times(1))
+        .getFeature(eq(requestPoint), Matchers.<StreamObserver<Feature>>any());
+    verify(client, times(1))
+        .info(
+            "Found feature called \"{0}\" at {1}, {2}",
```

Comment:
this is somewhat over specific isn't it?  We are checking that an RPC was made, not log statements.